AGENTS.md

```diff
@@ -0,0 +1,305 @@
+# AGENTS.md - Guidance for AI Tools
+
+This document provides essential information for AI coding assistants working with the OpenShift Kubernetes repository.
+
+## Critical Context: This is a Fork
+
+**This repository is OpenShift's fork of upstream Kubernetes (`kubernetes/kubernetes`).** The maintainers work to minimize the diff between upstream and this fork. AI tools must understand the special constraints and workflows that govern changes to this codebase.
+
+## Why This Matters for AI Tools
+
+Unlike most repositories where you can freely suggest changes anywhere in the codebase, this fork:
+
+1. **Gets rebased regularly** against upstream Kubernetes (every minor version release)
+2. **Must maintain minimal divergence** from upstream
+3. **Has strict commit message conventions** that affect how changes survive rebases
+4. **Requires changes to be categorized** as either upstream cherry-picks or downstream carries
+
+**Before suggesting any code changes, AI tools MUST understand the carry patch process documented below.**
+
+## Understanding the Rebase Process
+
+Every time a new minor Kubernetes version is released (e.g., v1.30.0 → v1.31.0), the maintainers perform a rebase with these steps:
+
+1. **Remove all carry patches** from the current branch
+2. **Cherry-pick all new upstream commits** added between the two minor releases
+3. **Reapply all carry patches** on top of the new upstream base
```

Comment:
We start with a clean branch from the upstream tag (e.g., `v1.31.0`) and cherry-pick all downstream carries from the current openshift/master branch. Then we merge our branch in `openshift/master`. The end result is `openshift/master` has the same code base as `v1.31.0` plus the downstream carries.